pkg/provision/storage/perWorkspaceStorage.go

```diff
@@ -0,0 +1,231 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	devfileConstants "github.com/devfile/devworkspace-operator/pkg/library/constants"
+	nsconfig "github.com/devfile/devworkspace-operator/pkg/provision/config"
+	"github.com/devfile/devworkspace-operator/pkg/provision/sync"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type perWorkspaceStorageProvisioner struct{}
+
+// TODO: Verify needsStorage is correct. Looks good but not sure.
+func (*perWorkspaceStorageProvisioner) NeedsStorage(workspace *dw.DevWorkspaceTemplateSpec) bool {
+	return needsStorage(workspace)
+}
+
+func syncPerWorkspacePVC(workspace *dw.DevWorkspace, clusterAPI sync.ClusterAPI) (*corev1.PersistentVolumeClaim, error) {
+	namespace := workspace.Namespace
+
+	namespacedConfig, err := nsconfig.ReadNamespacedConfig(namespace, clusterAPI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read namespace-specific configuration: %w", err)
+	}
+
+	// TODO: Should we adjust the storage size?
+	pvcSize := constants.PVCStorageSize
+	if namespacedConfig != nil && namespacedConfig.CommonPVCSize != "" {
+		pvcSize = namespacedConfig.CommonPVCSize
+	}
+
+	// We have to call clusterAPI.Client.Scheme() because the fakeClusterAPI.Scheme used in testing returns nil
+	pvc, err := getPerWorkspacePVCSpec(workspace, clusterAPI.Client.Scheme(), pvcSize)
+	if err != nil {
+		return nil, err
+	}
+
+	currObject, err := sync.SyncObjectWithCluster(pvc, clusterAPI)
+	switch t := err.(type) {
+	case nil:
+		break
+	case *sync.NotInSyncError:
+		return nil, &NotReadyError{
+			Message: "Updated per-workspace PVC on cluster",
+		}
+	case *sync.UnrecoverableSyncError:
+		return nil, &ProvisioningError{
+			Message: "Failed to sync per-workspace PVC to cluster",
+			Err:     t.Cause,
+		}
+	default:
+		return nil, err
+	}
+
+	currPVC, ok := currObject.(*corev1.PersistentVolumeClaim)
+	if !ok {
+		return nil, errors.New("tried to sync per-workspace PVC to cluster but did not get a PVC back")
+	}
+
+	return currPVC, nil
+
+}
+
+func getPerWorkspacePVCSpec(workspace *dw.DevWorkspace, scheme *runtime.Scheme, size string) (*corev1.PersistentVolumeClaim, error) {
```

Comment:
That sounds good to me, though one thing that came to mind:
`getPerWorkspacePVCSpec` also sets the controller reference, in addition to setting the PVC name.

Do you think it's better to keep a `getPerWorkspacePVCSpec` function that is just a wrapper for `getPVCSpec` (where it takes in the PVC name) but also sets up the controller reference?

Or just call `getPVCSpec` and then set up the controller reference immediately afterwards, whenever needed (both in the test method and in `syncPerWorkspacePVC`?

Reply:
> Do you think it's better to keep a `getPerWorkspacePVCSpec` function that is just a wrapper for `getPVCSpec` (where it takes in the PVC name) but also sets up the controller reference?

For now, I've done the above, though in theory it's not necessary to have a `getPerWorkspacePVCSpec` function as `provisionStorage` will get called, which should set up the controller reference when necessary. 

The issue I've found with relying on `provisionStorage` to set up the controller reference, is that certain test cases don't lead to the controller reference getting set. For example: when ephemeral storage is used, when storage is not required (`does-nothing-for-no-storage-needed.yaml`) and when the user specifies a project volume manually (`projects-volume-overriding.yaml`).

I was able to write some, kind of ugly workaround's in the test function so that the owner reference isn't checked for the above-mentioned cases. However, I haven't figured out yet how to detect the user-specified project volume case.

Something like:

```go
	ownerReferenceNotNeeded := false
	for _, component := range tt.Input.Workspace.Components {
		if component.Volume != nil && component.Volume.Ephemeral != nil {
			ownerReferenceNotNeeded = ownerReferenceNotNeeded || *component.Volume.Ephemeral
		}
	}

	ownerReferenceNotNeeded = ownerReferenceNotNeeded || !needsStorage(&workspace.Spec.Template)

	if !ownerReferenceNotNeeded {
		assert.NotEmpty(t, workspacePVC.ObjectMeta.OwnerReferences)
		assert.Len(t, workspacePVC.ObjectMeta.OwnerReferences, 1)
		assert.Equal(t, workspacePVC.ObjectMeta.OwnerReferences[0].Kind, "DevWorkspace")
	}

```
Should I go with this approach? I can update my PR with this approach if you would like to test things out.

Reply:
> Do you think it's better to keep a getPerWorkspacePVCSpec function that is just a wrapper for getPVCSpec (where it takes in the PVC name) but also sets up the controller reference?

In general it's better to not have functions optionally do anything (we don't want ownerrefs on commonPVC workspaces ever). This could be worked around by supplying the entire workspace to the `getSpecPVC` function, but it's simpler to just set the controllerRef afterwards (e.g. https://github.com/amisevsk/devworkspace-operator/commit/67f0a59c3d9f3e2817ef1fb614bb54deb0739d9f#diff-a3fc5cfc1fe497d9039f7540c20597bc9e87b4ebbe9ab58d6bc25b84dc647ed6R60-R63). This keeps the purpose of `getSpecPVC` simpler.

In the test method, you're calling `ProvisionStorage` so the PVC should get its ownerref set. To test that the PVC is created as expected, we should just be getting the PVC from the fake k8s client and checking its fields.